File: atcoder/tyama_atcodercodefestival2016qualcD.py
#!/usr/bin/python
import sys
if sys.version_info[0]<3: input=raw_input

N,M=map(int,input().split());C=[[0]*-~N for _ in range(N+1)];D=list(map(list,C));S=[input()for _ in range(N)];r=0
for t in range(M-1):
	for i in range(1,N+1):
		for j in range(1,N+1):C[i][j]=C[i-1][j-1]+int(S[i-1][t]==S[j-1][t+1]);D[i][j]=min(D[i-1][j],D[i][j-1])+C[i][j]
	r+=D[N][N]
print(r)

File: atcoder/codefestival/tyama_atcodercodefestival2016qualcD.rb
#!/usr/bin/ruby
N,M=gets.split.map &:to_i
C=(N+1).times.map{[0]*-~N}
D=C.map &:dup
S=*$<
r=0
(M-1).times{|t|
	1.upto(N){|i|1.upto(N){|j|
		C[i][j]=C[i-1][j-1]+(S[i-1][t]==S[j-1][t+1]?1:0)
		D[i][j]=[D[i-1][j],D[i][j-1]].min+C[i][j]
	}}
	r+=D[N][N]
}
p r
